examples/controlnet/train_controlnet_flux.py

```diff
@@ -0,0 +1,1321 @@
+        latent_image_ids = latent_image_ids.reshape(
+            batch_size, latent_image_id_height * latent_image_id_width, latent_image_id_channels
+        )
+
+        return latent_image_ids.to(device=device, dtype=dtype)
+
+    def _pack_latents(latents, batch_size, num_channels_latents, height, width):
+        latents = latents.view(batch_size, num_channels_latents, height // 2, 2, width // 2, 2)
+        latents = latents.permute(0, 2, 4, 1, 3, 5)
+        latents = latents.reshape(batch_size, (height // 2) * (width // 2), num_channels_latents * 4)
+
+        return latents
+
+    image_logs = None
+    for epoch in range(first_epoch, args.num_train_epochs):
+        for step, batch in enumerate(train_dataloader):
+            with accelerator.accumulate(flux_controlnet):
+                # Convert images to latent space
+
+                # vae encode
+                pixel_values = batch["pixel_values"].to(dtype=weight_dtype)
+                pixel_latents_tmp = vae.encode(pixel_values).latent_dist.sample()
+                pixel_latents_tmp = (pixel_latents_tmp - vae.config.shift_factor) * vae.config.scaling_factor
+                pixel_latents = _pack_latents(
+                    pixel_latents_tmp,
+                    pixel_values.shape[0],
+                    pixel_latents_tmp.shape[1],
+                    pixel_latents_tmp.shape[2],
+                    pixel_latents_tmp.shape[3],
+                )
+
+                control_values = batch["conditioning_pixel_values"].to(dtype=weight_dtype)
+                control_latents = vae.encode(control_values).latent_dist.sample()
+                control_latents = (control_latents - vae.config.shift_factor) * vae.config.scaling_factor
+                control_image = _pack_latents(
+                    control_latents,
+                    control_values.shape[0],
+                    control_latents.shape[1],
+                    control_latents.shape[2],
+                    control_latents.shape[3],
+                )
+
+                latent_image_ids = _prepare_latent_image_ids(
+                    batch_size=pixel_latents_tmp.shape[0],
+                    height=pixel_latents_tmp.shape[2],
+                    width=pixel_latents_tmp.shape[3],
+                    device=pixel_values.device,
+                    dtype=pixel_values.dtype,
+                )
+
+                # Sample noise that we'll add to the latents
+                noise = torch.randn_like(pixel_latents).to(accelerator.device).to(dtype=weight_dtype)
+                bsz = pixel_latents.shape[0]
+
+                # Sample a random timestep for each image
+                t = torch.sigmoid(torch.randn((bsz,), device=accelerator.device, dtype=weight_dtype))
+
+                # apply flow matching
+                noisy_latents = (
+                    1 - t.unsqueeze(1).unsqueeze(2).repeat(1, pixel_latents.shape[1], pixel_latents.shape[2])
```

Comment:
This feels a little too overwhelming. Could we match how we do it here?
https://github.com/huggingface/diffusers/blob/48e36353d8cbf0322ec1ad0684b95d11f70af2de/examples/dreambooth/train_dreambooth_lora_flux.py#L1647

Reply:
you mean remove the dim-expanding and repeating ops? i think it's better to write straight out those ops so make code readable and let reader know what the tensor shape is.